assembl/tasks/mac.py

```diff
@@ -27,80 +53,504 @@ def create_venv_python_3(c):
+            c.run("rm -f venv/bin/npm")  # remove the symlink first otherwise next command raises OSError: [Errno 17] File exists
+            c.run("nodeenv --node=%s --npm=%s --python-virtualenv assembl/static/js" % (n_version, npm_version))
+        upgrade_yarn_mac(c)
+        with c.cd(get_node_base_path(c)):
+            with venv(c):
+                c.run("npm install reinstall -g")
+
+        build.update_npm_requirements(force_reinstall=True)
+    else:
+        print "Node version OK"
+
+
+@task()
+def app_update_dependencies(c, force_reinstall=False, from_dump=False):
+    """
+    Updates all python and javascript dependencies.  Everything that requires a
+    network connection to update
+    """
+    if not from_dump:
+        ensure_requirements(c)
+    update_pip_requirements_mac(c, force_reinstall=force_reinstall)
+    update_node(c, force_reinstall=force_reinstall)
+    build.update_npm_requirements(c, force_reinstall=force_reinstall)
+
+
+@task()
+def app_reinstall_all_dependencies(c):
+    """
+    Reinstall all python and javascript dependencies.
+    Usefull after a OS upgrade, node upgrade, etc.
+    """
+    app_update_dependencies(c, force_reinstall=True)
+
+
+@task()
+def app_fullupdate(c):
+    """
+    Full Update: Update to latest git, update dependencies and compile app.
+    You need internet connectivity, and can't run this on a branch.
+    """
+    update_main_code(c)
+    create_local_ini(c)
+    app_compile(c)
+
+
+@task()
+def app_update(c):
+    """
+    Fast Update: Update to latest git, compile app but don't update requirements
+    Useful for deploying hotfixes.  You need internet connectivity, and can't
+    run this on a branch.
+    """
+    update_main_code(c)
+    app_compile_noupdate(c)
+
+
+@task()
+def app_compile(c):
+    """
+    Full Update: This is what you normally run after a git pull.
+    Doesn't touch git state, but updates requirements, rebuilds all
+    generated files annd restarts whatever needs restarting.
+    You need internet connectivity.  If you are on a plane, use
+    app_compile_noupdate instead.
+    """
+    app_update_dependencies(c)
+    app_compile_noupdate(c)
+
+
+@task()
+def app_compile_noupdate(c):
+    """
+    Fast Update: Doesn't touch git state, don't update requirements, and rebuild
+    all generated files. You normally do not need to have internet connectivity.
+    """
+    app_compile_nodbupdate(c)
+    app_db_update(c)
+    app_reload(c)
+
+
+def app_reload(c):
+    """
+    Restart all necessary processes after an update
+    """
+    if is_supervisor_running(c):
+        with venv(c):
+            c.run("supervisorctl stop dev:")
+            c.run("supervisorctl update")
+            c.run("supervisorctl restart dev:")
+
+
+@task()
+def app_db_update(c):
+    """
+    Migrates database using south
+    """
+    print('Migrating database')
+    with venv(c):
+        c.run('alembic -c %s upgrade head' % (c.config._internal.ini_file))
```

Comment:
This command should do a bit more guarding:
Check if the migration has more than 1 head, if it does, then alert on the console and bail out. Otherwise, continue with the line you have. If you can, use the API of alembic directly:
https://alembic.sqlalchemy.org/en/latest/api/commands.html#alembic.command.heads

If not, you can run:
`alembic -c local.ini history`

And check if the following pattern occurs more than once:
```
import re
regex = r"\w+ -> \w+ \(head\), .*"

# Imagine history was fetched from the alembic history command
history = """a3dc2d6f7562 -> 33735b0850fc (head), Create full text synthesis
464b02fb1006 -> a3dc2d6f7562 (head), Add share count field to idea model
3f239cc82989 -> 464b02fb1006, Add OldSlug table
96c1c41d223b -> 3f239cc82989, Add tables for tags association with posts
4f529b59026b -> 96c1c41d223b, Add summary to Annoucement"""

matches = re.findall(regex, history, re.MULTILINE)
if len(matches) > 1:
   raise Exception("You dun goofed, son")
else:
   c.run('alembic -c %s upgrade head' % (c.config._internal.ini_file))
```